dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -0,0 +1,574 @@
+
+    if (first_block_number < first_used_block)
+    {
+        window_blocks.erase(window_blocks.begin(),
+                            window_blocks.begin() + (first_used_block - first_block_number));
+        first_block_number = first_used_block;
+
+        assert(next_output_block_number >= first_block_number);
+        assert(frame_start.block >= first_block_number);
+        assert(prev_frame_start.block >= first_block_number);
+        assert(current_row.block >= first_block_number);
+    }
+}
+
+void WindowBlockInputStream::appendBlock(Block & current_block)
+{
+    if (!input_is_finished)
+    {
+        if (current_block.rows() == 0)
+        {
+            return;
+        }
+
+        window_blocks.push_back({});
+        auto & window_block = window_blocks.back();
+
+        window_block.rows = current_block.rows();
+
+        // Initialize output columns and add new columns to output block.
+        for (auto & ws : workspaces)
+        {
+            window_block.output_columns.push_back(ws.window_function->getReturnType()->createColumn());
+        }
+
+        window_block.input_columns = current_block.getColumns();
+    }
+
+    // Start the calculations. First, advance the partition end.
+    for (;;)
+    {
+        advancePartitionEnd();
+
+        // Either we ran out of data or we found the end of partition (maybe
+        // both, but this only happens at the total end of data).
+        assert(partition_ended || partition_end == blocksEnd());
+        if (partition_ended && partition_end == blocksEnd())
+        {
+            assert(input_is_finished);
+        }
+
+
+        while (current_row < partition_end)
+        {
+            // if window only have row_number function, we can ignore judging peers
+            if (!only_have_row_number)
+            {
+                // We now know that the current row is valid, so we can update the
+                // peer group start.
+                if (!arePeers(peer_group_start, current_row))
+                {
+                    peer_group_start = current_row;
+                    peer_group_start_row_number = current_row_number;
+                    ++peer_group_number;
+                }
+            }
+
+            // Advance the frame start.
+            advanceFrameStart();
+
+            if (!frame_started)
+            {
+                // Wait for more input data to find the start of frame.
+                assert(!input_is_finished);
+                assert(!partition_ended);
+                return;
+            }
+
+            // frame_end must be greater or equal than frame_start, so if the
+            // frame_start is already past the current frame_end, we can start
+            // from it to save us some work.
+            if (frame_end < frame_start)
+            {
+                frame_end = frame_start;
+            }
+
+            // Advance the frame end.
+            advanceFrameEnd();
+
+            if (!frame_ended)
```

Comment:
Might it not find `frame_end` when `partition_end` is already found?

Reply:
for agg function?

Reply:
> for agg function?

yes